data_science/archive_exploration/02 - exploring trees.ipynb

```diff
@@ -0,0 +1,353 @@
+    "We can obtain a little extra useful context from some of these records. The _arrangement_ of each collection is described in the `Arrangement` field of a collection-level record. However, these are inconsistently written and are unlikely to be machine-interpretable any time soon. Still, good to know that they're there"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "display(HTML(df['Arrangement'][269057][0]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's quickly clean up some of the json-ified data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df['AltRefNo'] = df['AltRefNo'].dropna().apply(lambda x: x[0])\n",
+    "df['Level'] = df['Level'].dropna().apply(lambda x: x[0])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### building a tree\n",
+    "Let's take a look at some of the collection-level records' `AltRefNo`s and use one of them as a starting point for our tree building"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "for collection_id in sorted(df['AltRefNo'][df['Level'] == 'Collection'].values):\n",
+    "    print(collection_id)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "alt_ref_no = 'PENROSE'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We want to find every record in the dataframe whose `AltRefNo`s start with the chosen string above:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "codes_as_str = sorted(df['AltRefNo'][df['AltRefNo']\n",
+    "                                     .str.startswith(alt_ref_no)\n",
+    "                                     .fillna(False)]\n",
+    "                      .tolist())\n",
+    "codes_as_str"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll now clean up the split those codes into lists of parts, using each `\\` as a delimiter."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "codes_as_list = sorted(list(map(lambda x: re.split('/|\\.', x.strip()), \n",
+    "                                codes_as_str)))\n",
+    "codes_as_list"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lets zip these strings and lists together into a dictionary - this is going to be useful in a minute..."
```

Comment:
sp: “Let’s”

Reply:
Are we guaranteed that `codes_as_list` has the same sort order as `codes_as_str`? And if so, why do we need to re-sort `codes_as_list`?

Also, I wonder if a dict comprehension might be neater here:

```python
codes = {
     c: re.split('/|\.', c.strip())
     for c in codes_as_str
}
```

You never use `codes_as_list` again.